### PR TITLE
Add modal component

### PR DESCRIPTION
**Pull request by emmabckstrm**
Example modal. Styles for overlay, modal window, close button, alignment of modal actions are implemented in this PR. Styling for modal content, in this case the heading and text, is up to implementation of each modal for flexibility. 

Screenshot 2020-09-01 at 13 50 28

Screenshot 2020-09-01 at 13 52 41

### CODE REVIEW

File: app/assets/main/components/modal.scss [hunk 1]
```diff
@@ -0,0 +1,36 @@
+.modal-container {
+  @apply fixed top-0 left-0 right-0 bottom-0 z-40
+    invisible;
+}
+
+.modal-overlay {
+  @apply absolute w-full h-full
+    bg-black opacity-75;
+}
```
Comment (emmabckstrm) [inline app/assets/main/components/modal.scss:9]: This is not optimal. I started with using background opacity (https://tailwindcss.com/docs/background-opacity), but it turns out we have not updated `tailwind`. When upgrading `tailwind`, this should be fixed

**Review by stefanlindbohm — APPROVED** [review]
Really like the clean markup to use this! And I guess it will be simpler yet with the new Tailwind helpers? 🙌 

One thing that sticks out a little to me is that the close button ("X") is not at the same padding that other elements are. Can we unify that so that it's in line with the top of the content and the right of the rightmost button? Or even make a  `modal-header` class that always goes at the top so that the close button can be vertically centered to the heading and right aligned with the rightmost button?

File: app/assets/main/controllers/modal_controller.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  open() {
+    this.containerTarget.style.visibility = 'visible';
+    this.contentTarget.querySelector('button').focus();
+  }
+
+  close() {
+    this.containerTarget.style.visibility = 'hidden';
+    this.triggerTarget.focus();
```
[the remark this answers is not in the document]
Comment (emmabckstrm) [reply app/assets/main/controllers/modal_controller.js:17]: We could change the name, but it's the trigger to open the modal and the reason for putting focus on it is accessibility reasons! We're not perfect in this dimension but I've made the whole modal possible to navigate with the keyboard @stefanlindbohm

File: app/assets/main/components/modal.scss (hunk 1 above)
Comment (emmabckstrm) [reply app/assets/main/components/modal.scss:9]: Nice! Will update in accordance

File: app/assets/main/controllers/modal_controller.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+export default class ModalController extends Controller {
+  initialize() {
+    const closeButton = this.createCloseButton();
+
+    this.insertElementFirst(closeButton, this.contentTarget);
+  }
+
+  open() {
+    this.containerTarget.style.visibility = 'visible';
+    this.contentTarget.querySelector('button').focus();
```
[the remark this answers is not in the document]
Comment (emmabckstrm) [reply app/assets/main/controllers/modal_controller.js:12]: nice suggestion!

**Comment by emmabckstrm** [comment]
@stefanlindbohm I totally hear you in regards to the close "x", however I still thinks it both looks cleaner (in a way) to keep it close to the corner and it's more "clickable" when it's close to the corner. Here are some screenshots, and to me the "x" looks more to be a part of the modal content in my honest opinion.  

Screenshot 2020-09-03 at 12 22 23
Screenshot 2020-09-03 at 12 22 28

File: app/assets/main/controllers/modal_controller.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  close() {
+    this.containerTarget.style.visibility = 'hidden';
+    this.triggerTarget.focus();
+  }
+
+  /* eslint-disable class-methods-use-this */
```
[the remark this answers is not in the document]
Comment (stefanlindbohm) [reply app/assets/main/controllers/modal_controller.js:20]: I mean make them static methods :). Too many programming languages!

File: app/assets/main/controllers/modal_controller.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefanlindbohm) [reply app/assets/main/controllers/modal_controller.js:17]: Oh, so it puts focus back on the button that opened the modal? That's very nice. Don't understand how, but it's nice :D

**Comment by stefanlindbohm** [comment]
> @stefanlindbohm I totally hear you in regards to the close "x", however I still thinks it both looks cleaner (in a way) to keep it close to the corner and it's more "clickable" when it's close to the corner. Here are some screenshots, and to me the "x" looks more to be a part of the modal content in my honest opinion.

Hmm, alright, I see your point. It's just that it looks really misaligned when the lines from header/button rows cross through the "X".

Come to think of it, shouldn't one of the buttons always be "cancel" and such the "X" button becomes kind of redundant? We could make clicking outside of the modal close it as a "panic close" :).

File: app/assets/main/controllers/modal_controller.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (emmabckstrm) [reply app/assets/main/controllers/modal_controller.js:17]: @stefanlindbohm Exactly! It's a really nice feature https://developer.mozilla.org/en-US/docs/Web/API/HTMLOrForeignElement/focus

**Comment by emmabckstrm** [comment]
@stefanlindbohm I don't think the close button is redundant as it's basically a universal symbol of closing a modal. Having a clear close button makes it redundant to read the text on the buttons :) 

If the close button is more aligned in smaller screens, are you ok then?

Screenshot 2020-09-03 at 13 44 09

**Comment by stefanlindbohm** [comment]
> @stefanlindbohm I don't think the close button is redundant as it's basically a universal symbol of closing a modal. Having a clear close button makes it redundant to read the text on the buttons :)
> 
> If the close button is more aligned in smaller screens, are you ok then? I personally don't mind the mis-alignment that much and rather have the functionality.
> 
> also this is just a random modal to exemplify, don't mind the content of the modal in this screenshot

Modal dialogs on Apple platforms don't have close buttons. But I saw that on Windows dialogs do have close buttons. I'm OK with having it now that it is more in line with other elements.

**Pull request merged**